### PR TITLE
Allow arbitrary data attributes in ``ExperimentResultData``

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Allows arbitrary data attributes in the `ExperimentResultData` class letting it function more like the dict container it is exposed to users as via the `Result.data` method.

### Details and comments

This (or something like this) will be needed if the future to allow Aer to add different data types to the result data container (such as probabilities and expectation values).

The existence of this class at all seems rather superfluous since it is always converted to a dict via the various `Result` methods like `data` `get_statevector` `get_unitary`. It could easily just be replaced by a plain dictionary in the `ExperimentResult` and `Result` classes and methods. I think it is only still here as a leftover from the Marshmallow classes.

### CODE REVIEW

File: qiskit/result/models.py [hunk 1]
```diff
@@ -22,35 +22,17 @@
 class ExperimentResultData:
     """Class representing experiment result data"""
 
-    def __init__(self, counts=None, snapshots=None, memory=None,
-                 statevector=None, unitary=None):
+    def __init__(self, **kwargs):
```
Comment (mtreinish) [inline qiskit/result/models.py:25]: Can we leave the old fields here and just add kwargs to the end? This will make it easier for existing users @jyu00

Comment (chriseclectic) [reply qiskit/result/models.py]: Done @mtreinish

**Review by mtreinish — COMMENTED** [review]
LGTM, one comment inline but not a big deal either way since this should work.  Before we merge this though it'd probably be good to have a test that exercises `to_dict` and `from_dict` with some other field like a snapshot or something just to validate it all works.

File: qiskit/result/models.py [hunk 2]
```diff
@@ -39,18 +39,27 @@ def __init__(self, counts=None, snapshots=None, memory=None,
...
+            self._data_attributes.append('memory')
             self.memory = memory
         if statevector is not None:
+            self._data_attributes.append('statevector')
             self.statevector = statevector
         if unitary is not None:
+            self._data_attributes.append('unitary')
             self.unitary = unitary
+        for key, value in kwargs.items():
+            setattr(self, key, value)
```
[the remark this answers is not in the document]
Comment (chriseclectic) [reply qiskit/result/models.py:61]: Yeah that would make sense, but is a bit beyond the scope of this PR.

**Review by mtreinish — APPROVED** [review]
This lgtm. The only thing I think is a potential issue is the `__repr__` doesn't take into account the extra kwargs.

**Review by mtreinish — APPROVED** [review]

**Pull request merged**